compositor|script: Update script's active_touch_points when hit tests on compositor fail

kongbai1996 opened the pull request:
[RootCause]:Tochdown executes normally, but touchup fails during hittest and does not send a touchup event to the script, causing the script to save an incorrect number of active_touch_points. The application receives an incorrect event.touches.length, causing a logic error and preventing the carousel from sliding. 
[Solution]:
When hit test on compositor fails, we also need to send the EmbedderToConstellationMessage::ForwardInputEvent message to constellation. In the script's handle_touch_event, if it exits early, we also need to update active_touch_points.

Testing: 
Fixes: #37763

kongbai1996 committed:
[RootCause]:
Tochdown executes normally, but touchup fails during hittest and does not send a touchup event to the script, causing the script to save an incorrect number of active_touch_points.
The application receives an incorrect event.touches.length, causing a logic error and preventing the carousel from sliding.
[Solution]:
We need to add a retry mechanism in the case of Err(HitTestError::Others).
When the retry fails, we also need to send the EmbedderToConstellationMessage::ForwardInputEvent message to constellation.
In the script's handle_touch_event, if it exits early, we also need to update active_touch_points.

Signed-off-by: kongbai1996 <1782765876@qq.com>

kongbai1996 commented on components/compositing/webview_renderer.rs:
```diff
@@ -370,15 +370,12 @@ impl WebViewRenderer {
         {
             Ok(hit_test_results) => hit_test_results,
-            Err(HitTestError::EpochMismatch) => {
+            Err(_) => {
```
There are some cases where there is no hittest result list, the specific reason has not been analyzed yet, I will analyze it later

kongbai1996 replied:
I checked this situation. When the problem occurs, the touch point information given by the web component of openharmony is incorrect, and the y value is negative, so there is no HitTestResult.

kongbai1996 replied:
The original logic is retained. Other errors are not retried. Only epoch mismatch is retried.

xiaochengh commented:
Could you reupload the diagram with a better resolution? I can't see the text in it...

kongbai1996 commented:
> 您能否以更好的分辨率重新上传图表？我看不到里面的文字......

Updated

kongbai1996 commented on components/script/dom/document.rs:
```diff
@@ -2354,6 +2356,34 @@ impl Document {
     }
 
+    // If hittest fails, we still need to update the active point information.
+    fn update_active_touch_points_when_early_return(&self, event: TouchEvent) {
```
A retry mechanism is still needed. Most of the time, retries are successful, and only a small portion of retries fail.

xiaochengh reviewed (commented):
Thanks for the updates! I think this is how we should move forward:
1. Fix the touch point information given from OHOS, since that seems to be the root cause of this bug. This should be done in a separate PR
2. In this PR, revert everything related to retry, and only let `script` reset active touch points when there are failed hit tests

Unconditionally retrying all hit tests isn't correct as the current mechanism is only supposed to work for epoch mismatches. So let's no do that.

kongbai1996 committed:
Do not change the logic of other hittest failures.
Extract the common logic of `dispatch_point_input_event` and `dispatch_pending_point_input_events` to the `dispatch_point_input_event_with_retry` method..

Signed-off-by: kongbai1996 <1782765876@qq.com>

kongbai1996 commented:
If we remove all the content related to retrying, there is a higher probability of hitting a test failure, which could result in a certain probability of the page not being able to slide with touch. I think we still need to retry once, and if the retry fails, then go to the logic of only updating the touch points.

xiaochengh commented on components/compositing/webview_renderer.rs:
```diff
@@ -389,15 +404,19 @@ impl WebViewRenderer {
-                self.global
-                    .borrow_mut()
-                    .update_cursor_from_hittest(point, &result);
+                if let Some(ref result) = result {
```
nit: Could you revert this block?

kongbai1996 replied:
Because I changed the type of `result` from `CompositorHitTestResul` to `Option<CompositorHitTestResult>`, a piece of logic is needed here to determine whether result has a value

xiaochengh commented on components/compositing/webview_renderer.rs:
```diff
@@ -345,7 +345,22 @@ impl WebViewRenderer {
+        }
+    }
+
+    pub(crate) fn dispatch_point_input_event_with_retry(
+        &self,
+        mut event: InputEvent,
+        need_retry: bool,
```
```suggestion
    pub(crate) fn dispatch_point_input_event_internal(
        &self,
        mut event: InputEvent,
        retry_on_error: bool,
```

kongbai1996 replied:
Modified

xiaochengh reviewed (approved):
LGTM with some small nits

xiaochengh commented:
Discussed with @kongbai1996 offline:
1. This PR will only update active touch status on failed hit tests, but not alter any retry logic
2. We may investigate and fix the root cause of hit test failures in a separate attempt

xiaochengh commented:
Please also modify the PR title and description, since this PR no longer modifies the retry mechanism

kongbai1996 committed:
rename common function name.

Signed-off-by: kongbai1996 <1782765876@qq.com>
